### DISCUSSION TITLE
How can I load the BIGG universal model with COBRApy?

### DISCUSSION THREAD

**Discussion by cdanielmachado** [opencobra/cobrapy#1296] [Q&A]
Trying to load the latest `universal_model.json` file from BiGG returns the following error:

```
~/opt/miniconda3/envs/py37/lib/python3.7/site-packages/cobra/core/object.py in annotation(self, annotation)
     48     def annotation(self, annotation):
     49         if not isinstance(annotation, dict):
---> 50             raise TypeError("Annotation must be a dict")
     51         else:
     52             self._annotation = annotation

TypeError: Annotation must be a dict
```

**Comment by cdiener** [comment]
Might be related to https://github.com/opencobra/cobrapy/issues/736 but would need a bit more diving into...

**Comment by Midnighter** [comment]
It's due to a more recent change https://github.com/opencobra/cobrapy/pull/930 where we moved to more consistent formats but once again didn't account for file format backward compatibility. :disappointed: 

I guess an annoying workaround would be to load the JSON with an older version of cobrapy, write it to SBML, and then load it again in the latest version if you need the latest version of cobrapy at all.

**Comment by zoey-rw** [comment]
The workaround (loading the JSON model with cobrapy v.0.17, but saving as SBML) isn't quite working for me. 
 It results in the error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-49d4759b12c1> in <module>
      9 for reaction in universal.reactions:
     10     reaction.annotation = {}
---> 11 cobra.io.write_sbml_model(universal, '/projectnb2/talbot-lab-data/metabolic_models/scripts/metabolic_model_curation/reference_data/universal_model.sbml')

~/.local/lib/python3.8/site-packages/cobra/io/sbml.py in write_sbml_model(cobra_model, filename, f_replace, **kwargs)
    798                     obj = metaid_map[member.getMetaIdRef()]
    799 
--> 800                 typecode = obj.getTypeCode()
    801                 obj_id = _check_required(obj, obj.getIdAttribute(), "id")
    802 

~/.local/lib/python3.8/site-packages/cobra/io/sbml.py in _model_to_sbml(cobra_model, f_replace, units)
    943         if "annotation" in meta:
    944             _sbase_annotations(doc, meta["annotation"])
--> 945         if "notes" in meta:
    946             _sbase_notes_dict(doc, meta["notes"])
    947 

~/.local/lib/python3.8/site-packages/cobra/io/sbml.py in _sbase_annotations(sbase, annotation)
   1336                  ["<p>{}: {}</p>".format(k, v) for (k, v) in notes.items()] + \
   1337                  ["</html>"]
-> 1338         _check(
   1339             sbase.setNotes("\n".join(tokens)),
   1340             "Setting notes on sbase: {}".format(sbase)

AttributeError: 'list' object has no attribute 'items'
```

Example script [here](https://github.com/zoey-rw/metabolic_model_curation/blob/main/universal_BiGG.ipynb)

**Comment by Midnighter** [comment]
In your notebook, you load the model successfully. Did you try to write it to SBML it without modifying the reactions' annotation attribute?

**Comment by zoey-rw** [comment]
No, after loading the model & before saving as SBML, I removed the annotation attributes as follows (lines 9-10 in the error message above):
```
for reaction in univ_bigg.reactions: 
    reaction.annotation = {}
```

**Comment by Midnighter** [comment]
Okay, I think we need to implement a better solution here. Probably, a wrapper that can upgrade old JSON versions to the newer format.

**Comment by cdiener** [comment]
Is BIGG still alive? It looks like it has been abandoned at this point... The easiest solution would still be for them to update the universal model...

**Comment by Midnighter** [comment]
No idea, if it's still alive tbh. It's probably not worth the effort but I feel personally responsible for the current JSON reading mess.

**Comment by cdiener** [comment]
I mean we could just convert the annotations in the JSON to the new standard and provide a repaired one then.

**Comment by cdiener** [comment]
As a workaround, I added a fixed version to [our wiki](https://github.com/opencobra/cobrapy/wiki/BIGG-Universal-Model).

**Comment by xinlia** [comment]
I also encountered this problem, so I wrote some code to modify the JSON file provided by BIGG. After testing, the output file can be loaded with Cobrapy.

```
def add_to_dir(my_dir, key, value):
    if key in my_dir:
        if isinstance(my_dir[key], list):
            my_dir[key].append(value)
        else:
            my_dir[key] = [my_dir[key], value]
    else:
        my_dir[key] = value


import json

with open('universal_model.json', 'r') as f:
    data = json.load(f)

for item in data['metabolites']+data['reactions']:
    annolist = item.get('annotation')
    if annolist:
        annodir={}
        for anno in annolist:
            annosplit = anno[1].split('/')
            key = annosplit[-2]
            value = annosplit[-1]
            add_to_dir(annodir, key, value)
        item['annotation'] = annodir
    else:
        item['annotation'] = {}

with open('universal_model_modified.json', 'w') as f:
    json.dump(data, f)
```

**Reply by xinlia** [reply]
I upload the modified model, if it will help you in some case.

[universal_model_modified.zip](https://github.com/opencobra/cobrapy/files/11365028/universal_model_modified.zip)

note: Original [universal_model.json](http://bigg.ucsd.edu/data_access) downloaded on May 1, 2023
